helix-core/src/main/java/org/apache/helix/task/TaskStateModelFactory.java

```diff
@@ -81,6 +79,12 @@ public Thread newThread(Runnable r) {
 
   @Override
   public TaskStateModel createNewStateModel(String resourceName, String partitionKey) {
+    if (_taskExecutor == null) {
+      int taskThreadPoolSize = getTaskThreadPoolSize();
+      updateLiveInstanceWithThreadPoolSize(taskThreadPoolSize);
```

Comment:
Curious about this design.
The createNewStateModel() is called when we have partition assigned, right? But I think the controller needs to know the pool size before assign the tasks. How it works?

Reply:
Answering "why lazy initialization" question here as well:

When `TaskStateModelFactory` is initialized, it is usually during factory registration. However, factory registration can occur before HelixManager connection, or after. When it occurs before HelixManager connection, `TaskStateModelFactory` actually cannot access the configs, therefore it cannot create the thread pool before manager connection. 

We have 2 ways to make it work: 1. we enforce `TaskStateModelFactory` initialization to be after HelixManager connection, 2. we lazily initialize the thread pool. I chose 2 because it is backward compatible - clients don't need to change their existing code in which they registered `TaskStateModelFactory` before HelixManager connection (which exists). 

In terms of `createNewStateModel()`, it is called during message handling when the state model of the message doesn't exist for the resource name and partition key. From the code I read, I don't think "the controller needs to know the pool size before assign the tasks", because as we can see from the original code of this class, the task thread pool is never exposed anywhere other than `createNewStateModel()`. In fact, it's not used anywhere other than `createNewStateModel()` either, therefore making it safe to initialize in this function. 

Feel free to correct me. @jiajunwang

Reply:
1. why we have to relies on the HelixManager to query for the config? You can use data accessor whenever it is requried. But if we don't even have the cluster/zk information, then it is not possible. This is one of the cons (or limitations) that we rely on ZK to pass the configuration.
2. The controller needs to know the real capacity otherwise how it ensures the assigned tasks fit the instance's capacity? I think this is the purpose of this design, no?
3. Alternatively, you can report the accepted capacity without really applying it to any executor. I mean, when liveinstance is created, put the capacity there. But at that moment, there is no real executor exists. Then when we create the executor, we apply for exactly the same number there.
To be clear, this is not a preferred design. In this way, the CURRENT_THREAD_POOL_SIZE becomes meaningless. But that is the only thing I can think of now.

Reply:
Reading your later comment, I think I know what you mean by "controller needs to know the pool size". Let me think about this.

Reply:
Resolved offline, thank you @jiajunwang!